src/internal/document_api/mod.rs

```diff
@@ -632,6 +637,39 @@ async fn resolve_keys_for_grants(
     ))
 }
 
+/// Get a cached policy or run the given Future to get the evaluated policy from the webservice.
+/// Policies that evaluate cleanly with no invalid users or groups are cached for future use.
+async fn get_cached_policy_or<F>(
+    config: &PolicyCachingConfig,
+    grant: &PolicyGrant,
+    policy_cache: &PolicyCache,
+    get_policy_f: F,
+) -> Result<(Vec<DocAccessEditErr>, Vec<WithKey<UserOrGroup>>), IronOxideErr>
+where
+    F: Future<Output = Result<PolicyResponse, IronOxideErr>>,
+{
+    // if there's a value in the cache, use it
+    if let Some(cached_policy) = policy_cache.get(grant) {
+        Ok((vec![], cached_policy.clone()))
+    } else {
+        // otherwise query the webservice and cache the result if there are no errors
+        match get_policy_f.await {
```

Comment:
Because you're just passing along the error in the `Err` case, this would be cleaned to me as a `map()` instead of a `match`.

Reply:
Absolutely.